torch/distributed/distributed_c10d.py

```diff
@@ -1064,6 +1064,159 @@ def recv(tensor, src=None, group=None, tag=0):
+}
+
+
+def adaptive_send(tensor, dst, group=None, tag=0):
+    """
+    Sends a tensor together with its `dtype` synchronously.
+
+    Args:
+        tensor (Tensor): Tensor to send.
+        dst (int): Destination rank.
+        group (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+        tag (int, optional): Tag to match send with remote recv
+
+    """
+    _check_single_tensor(tensor, "tensor")
+    if _rank_not_in_group(group):
+        _warn_not_in_group("send")
+        return
+
+    send_tensor_size=torch.tensor(tensor.size(),dtype=torch.int).to(tensor.device)
+    send_tensor_dtype=torch.tensor([Tensordtype2ncclDataType_t[str(tensor.dtype)]],dtype=torch.int).to(tensor.device)
+    # send_tensor_info[0:-1] contains the shape of `tensor`
+    # send_tensor_info[-1] is the dtype of `tensor`
+    send_tensor_info=torch.cat((send_tensor_size,send_tensor_dtype))
+    send_len_tensor_info=torch.tensor(send_tensor_info.size(),dtype=torch.int).to(tensor.device)
+
+    if group is None or group is GroupMember.WORLD:
+        default_pg = _get_default_group()
+        default_pg.send([send_len_tensor_info], dst, tag=tag-1).wait()
+        # send `tensor`'s `dtype` first
+        default_pg.send([send_tensor_info], dst, tag=tag+1).wait()
+        default_pg.send([tensor], dst, tag).wait()
+    else:
+        group_dst_rank = _get_group_rank(group, dst)
+        group.send([send_len_tensor_info], group_dst_rank, tag=tag-1).wait()
+        # send `tensor`'s `dtype` first
+        group.send([send_tensor_info], group_dst_rank, tag=tag+1).wait()
+        group.send([tensor], group_dst_rank, tag).wait()
+
+
+# Reverse mapping of `Tensordtype2ncclDataType_t`
+ncclDataType_t2Tensordtype={
+    "0": torch.int8,
+    "1": torch.uint8,
+    "2": torch.int32,
+
+    "4": torch.int64,
+
+    "6": torch.float16,
+    "7": torch.float32,
+    "8": torch.float64,
+    "9": torch.bfloat16,
+}
+
+
+def adaptive_recv(src=None, group=None, tag=0):
+    """
+    Receives a tensor with self-adaptive `dtype` synchronously.
+
+    Args:
+        src (int, optional): Source rank. Will receive from any
+            process if unspecified.
+        group (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+        tag (int, optional): Tag to match recv with remote send
+
+    Returns:
+        Sender rank, Received tensor
+        `Sender rank`==-1, if not part of the group
+        `Received tensor` is identical to its corresponding src tensor in `adaptive_send()`
+
+    """
+    if _rank_not_in_group(group):
+        _warn_not_in_group("recv")
+        return -1
+
+    if group is None:
+        pg = _get_default_group()
+    else:
+        pg = group
+
+    recv_len_tensor_info=torch.tensor([-1],dtype=torch.int).cuda(get_rank())
+    recv_tensor_dtype=torch.tensor([-1],dtype=torch.int).cuda(get_rank())
```

Comment:
Not sure why the tensor's residence has to be CUDA. What if there is no CUDA support? Same below.

Reply:
Also, `get_rank()` would return the rank of the current process in the entire group rather than a local CUDA device ID.